src/settings/settingsbase.h

```diff
@@ -0,0 +1,124 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef SETTINGSBASE_H
+#define SETTINGSBASE_H
+
+#include <QSettings>
+
+#include "loghandler.h"
+#include "setting.h"
+#include "settinggroup.h"
+
+/**
+ * @brief The SettingsBase is a singleton class that manages the underlying
+ * QtSetting storage object and serves the as the holder of all settings in the
+ * codebase dynamic or static. The lifetime of this singleton is tied to the
+ * lifetime of the QApplication.
+ *
+ * There are two types of settings currently in this codebase.
+ *
+ * * Static settings are settings defined at build time in a declarative manner
+ * e.g. the SETTING macro or the EXPERIMENTAL_FEATURE macro.
+ * * Dynamic settings are settings which are created on demand at runtime.
+ *
+ * All settings are created using the SettingsFactory, which creates a Setting
+ * object and registers it in the SettingsBase class.
+ *
+ * The only public constructor for Setting objects is the SettingsFactory
+ * constructor, therefore everytime a Setting object is created it is guaranteed
+ * to be registered in the settings base.
+ *
+ */
+class SettingsBase final : public QObject, public LogSerializer {
+  Q_OBJECT
+  Q_DISABLE_COPY_MOVE(SettingsBase)
+
+ public:
+  ~SettingsBase();
+
+  // LogSerializer interface
+  void serializeLogs(
+      std::function<void(const QString& name, const QString& logs)>&& callback)
+      override;
+
+  /**
+   * @brief Returns a registered setting based on their key.
+   *
+   * This will return a nullptr in case the setting is not registered.
+   *
+   * @param key
+   * @return Setting*
+   */
+  static Setting* getSetting(const QString& key) {
+    return instance()->m_registeredSettings.value(key);
+  }
+
+  static QString settingsFileName();
+
+  /**
+   * @brief Soft reset all registered settings.
+   *
+   * If the setting is configured to be removed on reset this will reset it,
+   * otherwise it will stay there.
+   *
+   */
+  static void reset();
```

Comment:
Is there any advantage in moving away from pattern used for other singletons, which use non-static methods and the caller obtains the singleton instance?

Reply:
No strong opinions, but I prefer this format because it is more restrictive than exposing the `instance`. For example, although `serializeLogs` is a public function, since `instance` is not public, only the logHandler which explicitly is given access to the instance can access it.

Reply:
A concern is that the cognitive load associated with reading and maintaining this code has increased, because it is different from the established singleton patterns in the code base.